File: packages/api/src/handlers/transactions.ts
import { GetCommand, PutCommand, QueryCommand, UpdateCommand } from '@aws-sdk/lib-dynamodb';
import type { APIGatewayProxyResultV2 } from 'aws-lambda';
import { docClient, getTableName, key } from '../db';

const tableName = () => getTableName();

export async function earn(
  tenantId: string,
  programId: string,
  body: { memberId: string; points: number; idempotencyKey?: string }
): Promise<APIGatewayProxyResultV2> {
  const { memberId, points, idempotencyKey } = body;
  if (!memberId || points == null || points < 0)
    return { statusCode: 400, body: JSON.stringify({ error: 'memberId and non-negative points required' }) };
  const pk = `TENANT#${tenantId}#PROGRAM#${programId}`;
  const skBalance = `BALANCE#${memberId}`;
  const txnId = idempotencyKey ?? `txn_${Date.now()}_${Math.random().toString(36).slice(2, 9)}`;
  const now = new Date().toISOString();
  const skTxn = `TXN#${now}#${txnId}`;

  await docClient.send(
    new UpdateCommand({
      TableName: tableName(),
      Key: { pk, sk: skBalance },
      UpdateExpression: 'SET points = if_not_exists(points, :zero) + :pts, updatedAt = :now',
      ExpressionAttributeValues: { ':zero': 0, ':pts': points, ':now': now },
    })
  );
  await docClient.send(
    new PutCommand({
      TableName: tableName(),
      Item: {
        pk,
        sk: skTxn,
        type: 'earn',
        memberId,
        points,
        idempotencyKey: idempotencyKey ?? null,
        createdAt: now,
        gsi1pk: `TENANT#${tenantId}`,
        gsi1sk: `TXN#${programId}#${now}#${txnId}`,
      },
    })
  );
  const getBalance = await docClient.send(new GetCommand({ TableName: tableName(), Key: { pk, sk: skBalance } }));
  const balance = getBalance.Item?.points ?? points;
  return {
    statusCode: 200,
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ transactionId: txnId, balance, points }),
  };
}

export async function burn(
  tenantId: string,
  programId: string,
  body: { memberId: string; points: number }
): Promise<APIGatewayProxyResultV2> {
  const { memberId, points } = body;
  if (!memberId || points == null || points < 0)
    return { statusCode: 400, body: JSON.stringify({ error: 'memberId and non-negative points required' }) };
  const pk = `TENANT#${tenantId}#PROGRAM#${programId}`;
  const skBalance = `BALANCE#${memberId}`;
  const res = await docClient.send(new GetCommand({ TableName: tableName(), Key: { pk, sk: skBalance } }));
  const current = res.Item?.points ?? 0;
  if (current < points)
    return { statusCode: 400, body: JSON.stringify({ error: 'Insufficient balance', balance: current }) };
  const now = new Date().toISOString();
  const txnId = `txn_${Date.now()}_${Math.random().toString(36).slice(2, 9)}`;
  const skTxn = `TXN#${now}#${txnId}`;

  await docClient.send(
    new UpdateCommand({
      TableName: tableName(),
      Key: { pk, sk: skBalance },
      UpdateExpression: 'SET points = points - :pts, updatedAt = :now',
      ExpressionAttributeValues: { ':pts': points, ':now': now },
    })
  );
  await docClient.send(
    new PutCommand({
      TableName: tableName(),
      Item: {
        pk,
        sk: skTxn,
        type: 'burn',
        memberId,
        points,
        createdAt: now,
        gsi1pk: `TENANT#${tenantId}`,
        gsi1sk: `TXN#${programId}#${now}#${txnId}`,
      },
    })
  );
  const newBalance = current - points;
  return {
    statusCode: 200,
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ transactionId: txnId, balance: newBalance, points }),
  };
}

export async function getBalance(
  tenantId: string,
  programId: string,
  memberId: string
): Promise<APIGatewayProxyResultV2> {
  const pk = `TENANT#${tenantId}#PROGRAM#${programId}`;
  const sk = `BALANCE#${memberId}`;
  const res = await docClient.send(new GetCommand({ TableName: tableName(), Key: { pk, sk } }));
  const points = res.Item?.points ?? 0;
  return {
    statusCode: 200,
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ memberId, programId, balance: points }),
  };
}

File: packages/backend/src/app/services/transactions.py
"""Earn, burn, balance, and list transactions."""

import time
import random
import string
from botocore.exceptions import ClientError
from app.db import get_table, get_doc_client, get_table_name, serialize_item, key
from app.exceptions import BadRequestError, ConflictError

# Transaction items expire after 18 months (DynamoDB TTL attribute)
_TXN_TTL_SECONDS = 18 * 30 * 24 * 3600

# Idempotency records expire after 7 days
_IDEM_TTL_SECONDS = 7 * 24 * 3600


def _table():
    return get_table()


def _txn_id() -> str:
    return f"txn_{int(time.time()*1000)}_{''.join(random.choices(string.ascii_lowercase + string.digits, k=9))}"


def _txn_ttl() -> int:
    return int(time.time()) + _TXN_TTL_SECONDS


def _idem_ttl() -> int:
    return int(time.time()) + _IDEM_TTL_SECONDS


def earn(tenant_id: str, program_id: str, member_id: str, points: int, idempotency_key: str | None = None) -> dict:
    if points < 0:
        raise BadRequestError("memberId and non-negative points required")
    pk = key.program_scoped_pk(tenant_id, program_id)
    sk_balance = key.balance_sk(member_id)
    txn_id = idempotency_key or _txn_id()
    now = time.strftime("%Y-%m-%dT%H:%M:%S.000Z", time.gmtime())
    sk_txn = key.txn_sk(now, txn_id)
    table_name = get_table_name()
    tbl = _table()

    # Build atomic transaction: balance update + transaction record
    items = [
        {
            "Update": {
                "TableName": table_name,
                "Key": {"pk": {"S": pk}, "sk": {"S": sk_balance}},
                "UpdateExpression": "SET points = if_not_exists(points, :zero) + :pts, updatedAt = :now",
                "ExpressionAttributeValues": {
                    ":zero": {"N": "0"},
                    ":pts": {"N": str(points)},
                    ":now": {"S": now},
                },
            }
        },
        {
            "Put": {
                "TableName": table_name,
                "Item": serialize_item({
                    "pk": pk,
                    "sk": sk_txn,
                    "type": "earn",
                    "memberId": member_id,
                    "points": points,
                    "createdAt": now,
                    "ttl": _txn_ttl(),
                    "gsi1pk": key.gsi1_tenant(tenant_id),
                    "gsi1sk": key.gsi1_txn_sk(program_id, now, txn_id),
                }),
            }
        },
    ]

    # Idempotency guard: IDEM record with ConditionExpression rejects duplicates
    if idempotency_key:
        items.append({
            "Put": {
                "TableName": table_name,
                "Item": serialize_item({
                    "pk": pk,
                    "sk": key.idem_sk(idempotency_key),
                    "type": "earn",
                    "memberId": member_id,
                    "points": points,
                    "createdAt": now,
                    "ttl": _idem_ttl(),
                }),
                "ConditionExpression": "attribute_not_exists(sk)",
            }
        })

    try:
        get_doc_client().transact_write_items(TransactItems=items)
    except ClientError as e:
        if e.response["Error"]["Code"] == "TransactionCanceledException" and idempotency_key:
            # Duplicate detected — look up existing IDEM record
            existing = tbl.get_item(Key={"pk": pk, "sk": key.idem_sk(idempotency_key)})
            item = existing.get("Item")
            if item:
                if item.get("memberId") != member_id or int(item.get("points", -1)) != points:
                    raise ConflictError("Idempotency key already used with different parameters")
                # Return cached response (balance unchanged since earn was not re-applied)
                cached_balance = item.get("balance")
                if cached_balance is not None:
                    return {"transactionId": idempotency_key, "balance": int(cached_balance), "points": points}
                bal_res = tbl.get_item(Key={"pk": pk, "sk": sk_balance})
                return {"transactionId": idempotency_key, "balance": int(bal_res.get("Item", {}).get("points", 0)), "points": points}
        raise

    # Read updated balance
    res = tbl.get_item(Key={"pk": pk, "sk": sk_balance})
    balance = int(res.get("Item", {}).get("points", points))

    # Store balance in IDEM record for future duplicate responses (fire-and-forget)
    if idempotency_key:
        try:
            tbl.update_item(
                Key={"pk": pk, "sk": key.idem_sk(idempotency_key)},
                UpdateExpression="SET balance = :bal",
                ExpressionAttributeValues={":bal": balance},
            )
        except Exception:
            pass

    return {"transactionId": txn_id, "balance": balance, "points": points}


def burn(tenant_id: str, program_id: str, member_id: str, points: int) -> dict:
    if points < 0:
        raise BadRequestError("memberId and non-negative points required")
    pk = key.program_scoped_pk(tenant_id, program_id)
    sk_balance = key.balance_sk(member_id)
    tbl = _table()

    # Early check for better error message (not the real guard)
    res = tbl.get_item(Key={"pk": pk, "sk": sk_balance})
    current = int(res.get("Item", {}).get("points", 0))
    if current < points:
        raise BadRequestError(f"Insufficient balance (balance={current})")

    now = time.strftime("%Y-%m-%dT%H:%M:%S.000Z", time.gmtime())
    txn_id = _txn_id()
    sk_txn = key.txn_sk(now, txn_id)
    table_name = get_table_name()

    try:
        get_doc_client().transact_write_items(TransactItems=[
            {
                "Update": {
                    "TableName": table_name,
                    "Key": {"pk": {"S": pk}, "sk": {"S": sk_balance}},
                    "UpdateExpression": "SET points = points - :pts, updatedAt = :now",
                    "ConditionExpression": "points >= :pts",
                    "ExpressionAttributeValues": {
                        ":pts": {"N": str(points)},
                        ":now": {"S": now},
                    },
                }
            },
            {
                "Put": {
                    "TableName": table_name,
                    "Item": serialize_item({
                        "pk": pk,
                        "sk": sk_txn,
                        "type": "burn",
                        "memberId": member_id,
                        "points": points,
                        "createdAt": now,
                        "ttl": _txn_ttl(),
                        "gsi1pk": key.gsi1_tenant(tenant_id),
                        "gsi1sk": key.gsi1_txn_sk(program_id, now, txn_id),
                    }),
                }
            },
        ])
    except ClientError as e:
        if e.response["Error"]["Code"] == "TransactionCanceledException":
            raise BadRequestError("Insufficient balance")
        raise

    return {"transactionId": txn_id, "balance": current - points, "points": points}


def get_balance(tenant_id: str, program_id: str, member_id: str) -> dict:
    pk = key.program_scoped_pk(tenant_id, program_id)
    sk = key.balance_sk(member_id)
    res = _table().get_item(Key={"pk": pk, "sk": sk})
    points = res.get("Item", {}).get("points", 0)
    return {"memberId": member_id, "programId": program_id, "balance": points}


def list_transactions(
    tenant_id: str,
    program_id: str,
    member_id: str | None = None,
    limit: int = 50,
    next_token: str | None = None,
) -> tuple[list[dict], str | None]:
    pk = key.program_scoped_pk(tenant_id, program_id)
    params = {
        "KeyConditionExpression": "pk = :pk AND begins_with(sk, :sk)",
        "ExpressionAttributeValues": {":pk": pk, ":sk": key.TXN_SK_PREFIX},
        "Limit": min(limit, 100),
        "ScanIndexForward": False,
    }
    if next_token:
        params["ExclusiveStartKey"] = {"pk": pk, "sk": next_token}
    result = _table().query(**params)
    items = result.get("Items", [])
    next_key = result.get("LastEvaluatedKey")
    next_sk = next_key.get("sk") if next_key else None
    out = []
    for it in items:
        if member_id and it.get("memberId") != member_id:
            continue
        sk = it.get("sk", "")
        txn_id = sk.split("#")[-1] if "#" in sk else ""
        out.append({
            "transactionId": txn_id,
            "type": it.get("type", ""),
            "memberId": it.get("memberId", ""),
            "points": it.get("points", 0),
            "rewardId": it.get("rewardId"),
            "createdAt": it.get("createdAt", ""),
        })
    return out, next_sk
